File: apps/expo-example/src/toolkit.kt
// Implements the `@Native` declarations in toolkit.lucent.ts. The generated
// package declaration is prepended when this file omits one.

fun lucentNative_sha256(bytes: ArrayBuffer): String =
  java.security.MessageDigest.getInstance("SHA-256").digest(LucentBytes.toByteArray(bytes)).joinToString("") {
    (it.toInt() and 255).toString(16).padStart(2, '0')
  }

fun lucentNative_temporaryDirectory(): String =
  System.getProperty("java.io.tmpdir") ?: throw LucentError("FILE_TEMP", "Temporary directory unavailable")

fun lucentNative_writeFile(path: String, bytes: ArrayBuffer) {
  try {
    java.io.File(path).writeBytes(LucentBytes.toByteArray(bytes))
  } catch (error: java.io.IOException) {
    throw LucentError("FILE_WRITE", "Unable to write file", mapOf("path" to path))
  }
}

fun lucentNative_readFile(path: String): ArrayBuffer {
  try {
    return LucentBytes.fromByteArray(java.io.File(path).readBytes())
  } catch (error: java.io.IOException) {
    throw LucentError("FILE_READ", "Unable to read file", mapOf("path" to path))
  }
}

fun lucentNative_deviceModel(): String = android.os.Build.MODEL

File: apps/expo-example/src/toolkit.swift
import Foundation
import CryptoKit
import UIKit

// Implements the `@Native` declarations in toolkit.lucent.ts. Each function is
// `throws` because the generated caller marks every native call with `try`.

func lucentNative_sha256(bytes: ArrayBuffer) throws -> String {
  return SHA256.hash(data: LucentBytes.data(bytes)).map { String(format: "%02x", $0) }.joined()
}

func lucentNative_temporaryDirectory() throws -> String {
  return NSTemporaryDirectory()
}

func lucentNative_writeFile(path: String, bytes: ArrayBuffer) throws -> Void {
  do {
    try LucentBytes.data(bytes).write(to: URL(fileURLWithPath: path), options: .atomic)
  } catch {
    throw LucentError(code: "FILE_WRITE", message: "Unable to write file", metadata: ["path": path])
  }
}

func lucentNative_readFile(path: String) throws -> ArrayBuffer {
  do {
    return try LucentBytes.fromData(Data(contentsOf: URL(fileURLWithPath: path)))
  } catch {
    throw LucentError(code: "FILE_READ", message: "Unable to read file", metadata: ["path": path])
  }
}

@MainActor
func lucentNative_deviceModel() throws -> String {
  return UIDevice.current.model
}
